### PR TITLE
[WORKOUTGEN-4] Refactor + New features

### PR DESCRIPTION
**Pull request by oliviakschotz**
* Refactors view logic
* Creates new Preview screen where user can preview the generated workout before saving
* Adds ability to replace individual exercise with a new randomly generated one on preview screen (not in saved workouts tab)

# Demo
<video src="https://github.com/sallison921/workout-generator/assets/57269931/64e315ff-48c8-43e4-a529-b2efd7ffb82f" height=500></video>

### CODE REVIEW

File: WorkoutGenerator/Assets.xcassets/Loading.dataset/Animation - 1716329865815.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{"assets":[],"layers":[{"ddd":0,"ind":0,"ty":4,"nm":"Shape Layer 1","ks":{"o":{"k":100},"r":{"k":0},"p":{"k":[300,300,0]},"a":{"k":[0,0,0]},"s":{"k":[244,244,100]}},"ao":0,"shapes":[{"d":1,"ty":"el","s":{"k":[100,100]},"p":{"k":[0,0]},"nm":"Ellipse Path 1","mn":"ADBE Vector Shape - Ellipse"},{"ty":"tm","s":{"k":[{"i":{"x":[0.439],"y":[1.016]},"o":{"x":[0.561],"y":[0.015]},"n":["0p439_1p016_0p561_0p015"],"t":5,"s":[100],"e":[0]},{"t":33.0000013441176}],"ix":1},"e":{"k":[{"i":{"x":[0.439],"y":[1.017]},"o":{"x":[0.561],"y":[0.016]},"n":["0p439_1p017_0p561_0p016"],"t":0,"s":[100],"e":[0]},{"t":30.0000012219251}],"ix":2},"o":{"k":0,"ix":3},"m":1,"ix":2,"nm":"Trim Paths 1","mn":"ADBE Vector Filter - Trim"},{"ty":"st","fillEnabled":true,"c":{"k":[1,0.5843,0,1]},"o":{"k":100},"w":{"k":16},"lc":2,"lj":1,"ml":4,"nm":"Stroke 1","mn":"ADBE Vector Graphic - Stroke"}],"ip":0,"op":900.000036657751,"st":0,"bm":0,"sr":1}],"v":"4.5.3","ddd":0,"ip":1.00000004073083,"op":33.0000013441175,"fr":29.9700012207031,"w":600,"h":600}
```
Comment (oliviakschotz) [inline WorkoutGenerator/Assets.xcassets/Loading.dataset/Animation - 1716329865815.json:1]: New lottie file for a loading animation

File: WorkoutGenerator.xcodeproj/project.pbxproj
Comment (oliviakschotz) [inline WorkoutGenerator.xcodeproj/project.pbxproj:1]: Added lottie package for animation assets

File: WorkoutGenerator/PreviewWorkoutView.swift [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+//  WorkoutGenerator
+//
+//  Created by Olivia Schotz on 5/21/24.
+//
+
+import Foundation
+import SwiftUI
+import Lottie
+
+struct PreviewWorkoutView: View {
```
Comment (oliviakschotz) [inline WorkoutGenerator/PreviewWorkoutView.swift:14]: new preview view so users can see the proposed generated workout before saving

File: WorkoutGenerator/SavedWorkouts.swift [hunk 3]
```diff
@@ -9,155 +9,31 @@ import SwiftUI
...
-    
-
 }
-        
-        
-        
-        
-        
-        
-//    @State private var chosenItem: ExerciseType?
```
Comment (oliviakschotz) [inline WorkoutGenerator/SavedWorkouts.swift:114]: oh i deleted this for my branch but can put back for main

File: WorkoutGenerator/GeneratePlan.swift [hunk 4]
```diff
@@ -14,111 +14,36 @@ struct GeneratePlan: View {
     let encoder = JSONEncoder()
     let defaults = UserDefaults.standard
     
+    @State var selectedType: WorkoutType = .pull
```
Comment (sallison921) [inline WorkoutGenerator/GeneratePlan.swift:17]:
```suggestion
    @State private var selectedType: WorkoutType = .pull
```

File: WorkoutGenerator/GeneratePlan.swift [hunk 5]
```diff
@@ -14,111 +14,36 @@ struct GeneratePlan: View {
...
+    private func generateWorkoutButtonView(for workout: WorkoutType) -> some View {
+        NavigationLink(
+            destination: { PreviewWorkoutView(workoutType: workout) },
+            label: {
+                Text(workout.rawValue)
+                    .frame(width: 100, height: 100)
+                    .foregroundColor(.black)
+                    .background(Color(cgColor: .init(gray: 0.8, alpha: 1)))
+            })
+    }
```
Comment (sallison921) [inline WorkoutGenerator/GeneratePlan.swift:42]: should be:
```
 NavigationLink {
          PreviewWorkoutView(workoutType: workout) 
} label: {
                Text(workout.rawValue)
                    .frame(width: 100, height: 100)
                    .foregroundStyle.black)
                    .background(Color(cgColor: .init(gray: 0.8, alpha: 1)))
 }

```

File: WorkoutGenerator/GeneratePlan.swift [hunk 6]
```diff
@@ -14,111 +14,36 @@ struct GeneratePlan: View {
...
+    
+    func saveWorkout(workout: Workout, key: String) -> Void {
         
-        if let encoded = try? encoder.encode(workoutType) {
+        if let encoded = try? encoder.encode(workout) {
             defaults.set(encoded, forKey: key)
         } else {
             print("no")
```
Comment (sallison921) [inline WorkoutGenerator/GeneratePlan.swift:49]: we need to handle this gracefully hahah I forgot I did that - we could `fatalError()` for now so we know for REAL when it happens lol

File: WorkoutGenerator/PreviewWorkoutView.swift (hunk 2 above)
Comment (sallison921) [reply WorkoutGenerator/PreviewWorkoutView.swift:14]: sweet

File: WorkoutGenerator/PreviewWorkoutView.swift [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    
+    /// the type of workout to be generated
+    let workoutType: WorkoutType
+    /// the generated workout
+    @State var workout: Workout?
+    
+    /// Indicates which exercise the user is editing
+    @State var selectedExerciseIndex: Int? = nil
+    /// Used as a binding to show/hide the bottom sheet
+    @State var showBottomSheet: Bool = false
```
Comment (sallison921) [inline WorkoutGenerator/PreviewWorkoutView.swift:31]: private vars unless they're needed to be public?

File: WorkoutGenerator/SavedWorkouts.swift [hunk 8]
```diff
@@ -9,155 +9,31 @@ import SwiftUI
...
-//        List(model.workoutList, id: \.self) { workout in
-//            //https://forums.swift.org/t/concrete-type-conformance-to-hashable-protocol/20443/5
-//                //may need to use enums ugh
-//        }
-        List {
-            Section("Quad Workout"){
-                ForEach(savedQuadExercises, id: \.self){ item in
+        List(Array(workouts.keys), id: \.self) { workoutType in
+            Section("\(workoutType.rawValue) Workout"){
+                ForEach(workouts[workoutType] ?? [], id: \.self){ item in
```
Comment (sallison921) [inline WorkoutGenerator/SavedWorkouts.swift:19]: let's make Exercise Identifiable so we don't need to deal with the `id` part

**Review by sallison921 — COMMENTED** [review]
I mostly just put nits because I noticed them - excited about the animations!

**Pull request merged**